Initial contribution of smarthome packaging sample

chrisschauer opened the pull request:
See: https://github.com/eclipse/smarthome/issues/602

Commit:
Initial contribution of smarthome packaging sample

Signed-off-by: Christoph Schauer <christoph@schauer-itconsulting.de>

kaikreuzer commented on README.md:
```diff
@@ -1,3 +1,108 @@
-# Eclipse SmartHome Packaging Sample
+# Eclipse Smarthome Packaging Sample
```
a capital "H" please

kaikreuzer commented on README.md:
```diff
@@ -1,3 +1,108 @@
+
+More information about concierge: https://www.eclipse.org/concierge/index.php
+
+1. Prerequisites - Install Maven
```
Also mention that Java8 is required (the runtime will throw exceptions, if you try to start it on Java7 or less).

JochenHiller replied:
I had exactly this issue as JavaSE 7 was my default JVM.
@kaikreuzer should this maybe even be checked in start script? (start.sh)

chrisschauer replied:
Version is now checked in start.sh

kaikreuzer commented on README.md:
````diff
@@ -1,3 +1,108 @@
+Extract the distribution zip file and start the runtime:
+```
+unzip smarthome-packaging-sample-[version].zip
+cd runtime/concierge
````
would it be possible to move the `start.sh` script to the root, so that this line is not needed? Or create a new `start.sh`in the root that does nothing else than cd'ing and launching the other `start.sh`.

JochenHiller replied:
I would recommend the same

kaikreuzer commented on README.md:
````diff
@@ -1,3 +1,108 @@
+## How to add further bundles?
+1. Add your bundle as dependecy in the pom.xml. For example:
+```
+  <!-- Eclipse Smarthome dependencies - Bindings -->
````
capital "H" :-)

kaikreuzer commented on README.md:
````diff
@@ -1,3 +1,108 @@
+```
+2. Add the install and startup command to XARGS file. For example:
+```
+# Eclipse Smarthome Bindings. Start all bindings here
````
And "H" at all other occurrences as well, please!

kaikreuzer reviewed (approved):
This looks awesome and it worked for me right away - well done!
Besides a few minor comments, I'd be fine to merge it.

kaikreuzer commented:
@jochenhiller, Do you maybe want to have a look as well to see if the Concierge usage/configuration is according to the best practices or if you have any suggestions on what should be changed?

JochenHiller commented:
@kaikreuzer Yep, I will also make a review if this PR

JochenHiller commented on distro/runtime/concierge/smarthome.xargs:
```diff
@@ -0,0 +1,236 @@
+# -istart ${system.dir}/org.apache.felix.gogo.shell-0.*.jar
+ 
+# Concierge extensions, services (fragment, do NOT start)
+#-install ${concierge.dir}/org.eclipse.concierge.extension.permission-${concierge.version}*.jar
```
I would remove that line. If needed only include the OSGi permission classes as service bundle, and not as framework extension

JochenHiller commented on distro/runtime/concierge/smarthome.xargs:
```diff
@@ -0,0 +1,236 @@
+ 
+# Concierge extensions, services (fragment, do NOT start)
+#-install ${concierge.dir}/org.eclipse.concierge.extension.permission-${concierge.version}*.jar
+-istart ${concierge.dir}/org.eclipse.concierge.service.permission-${concierge.version}*.jar
```
Are your sure the this permission bundle will really be needed?
I have in mind that the Equinox console needed that. I am not aware of other bundles

chrisschauer replied:
No, I'm not sure. With extension.permission the runtime does not come up! The runtime and all bundles starts with and without the permission bundle. But i have no clue what happens in "action". I thought it belongs to the runtime, so i had to start it.

JochenHiller commented on distro/runtime/etc/services.cfg:
```diff
@@ -0,0 +1,34 @@
+org.eclipse.smarthome.threadpool:discovery=3
+
+# Non-scheduled thread pools can also provide a max size
+org.eclipse.smarthome.threadpool:safeCall=3,10
```
In this file there are some configs of bindings which may not be installed.
I think it makes sense to have them in, even if not use. So just copy binding to addons is sufficient.
It follows same pattern as used in openHAB2 distribution.
Good idea !

JochenHiller commented on distro/runtime/concierge/start.sh:
```diff
@@ -0,0 +1,45 @@
+
+# Find the concierge framework jar
+MAIN=$(find framework -name "org.eclipse.concierge-5.0.0*.jar" | sort | tail -1);
+
```
Check here first if we are running on (at least) JavaSE 8?

chrisschauer replied:
O.k. there is now a check on the version. If a JDK smaller than 8 is used the following will displayed:

```
Using java: Q:\software\jdk1.7.0_75
java version "1.7.0_75"
Java(TM) SE Runtime Environment (build 1.7.0_75-b13)
Java HotSpot(TM) 64-Bit Server VM (build 24.75-b04, mixed mode)
ERROR: JVM must be greater than 1.7
```

JochenHiller reviewed (changes requested):
Looks good overall !!! Thanks for this contribution !!!

Commit:
Added a capital "H" for every SmartHome occurrence. Added hint for using JDK8.

Commit:
Added a capital "H" for every SmartHome occurrence. Remove comment line from XARGs file.

Commit:
Added java version check before launching the runtime.

Commit:
Added start.sh in the distribution root folder

Commit:
Updated Readme

chrisschauer commented:
Hi @kaikreuzer, @JochenHiller 

i updated the PR. 
 * Added capital "H" on every occurrence
 * Added a version check on start.sh
 * Added a hint on readme to use correct java version
 * Added a start.sh on distribution root (only refers to the start.sh in the concierge folder)

Could you please check again! Thx

maggu2810 commented on distro/runtime/concierge/smarthome.xargs:
```diff
@@ -0,0 +1,235 @@
+# jUPNP
+-istart ${system.dir}/org.jupnp-2.2.0*.jar
+
+# jMDNS
```
JmDNS

maggu2810 commented on distro/runtime/concierge/smarthome.xargs:
```diff
@@ -0,0 +1,235 @@
+# Persistence: MapDB
+-istart ${system.dir}/mapdb-1.0.9*.jar
+
+# jUPNP
```
jUPnP

maggu2810 commented on distro/runtime/concierge/start.sh:
```diff
@@ -0,0 +1,51 @@
+	-Dfelix.fileinstall.active.level=4 \
+	-Djava.awt.headless=true \
+	-Dfile.encoding="UTF-8" \
+	-jar $MAIN $DIRNAME/smarthome.xargs
```
You should use quotes if you use environment variables for the values of `-D` to handle spaces correctly (not for $JAVA_OPTS).

maggu2810 reviewed (commented):
Some minor comments.

Commit:
Fixed name of 3rd party libs in xargs file. Use quotes on -D parameter to define paths with spaces.

chrisschauer commented:
Hi @maggu2810. I included your minor comments.

Commit:
Added hint how to use UIs. Added hint because there is no modeling in this distribution.

Commit:
Fixed typo.

JochenHiller commented:
This contribution looks really good now.
Thanks Chris for your work !!!

kaikreuzer commented:
Great, thank you very much!

chrisschauer commented:
You're welcome :-)
Thx @JochenHiller for the preparation work. Thx @kaikreuzer for the support

kaikreuzer commented:
https://twitter.com/kaikreuzer/status/780746179214774272 :-)
